### PR TITLE
Fix Xcode 10 errors relating to third-party

### LINKED ISSUE #20774: When new Xcode build system is used, Xcode build fails to download third-party dependencies
# Workaround (edit by @hramos)

You may work around this by manually installing third party dependencies into your project. The following steps are based on the [workaround provided](https://github.com/facebook/react-native/issues/14382#issuecomment-420190830) by @lyon007 in #14382.

1. ` rm -rf node_modules/ && yarn cache clean && yarn install`  and ` rm -rf ~/.rncache`
2. ` cd node_modules/react-native/scripts` 
3.  `./ios-install-third-party.sh`
      If your network is slow, you can download the packages separately and move them to the `rncache ` folder
      _1_. [folly-2016.10.31.00](https://codeload.github.com/facebook/folly/tar.gz/v2016.10.31.00)
      _2_.  [double-conversion-1.1.5](https://codeload.github.com/google/double-conversion/tar.gz/v1.1.5)
      _3_.  [glog-0.3.5 ](https://codeload.github.com/google/glog/tar.gz/v0.3.5)
      _4_.  [boost_1_63_0](https://github.com/react-native-community/boost-for-react-native/releases/download/v1.63.0-0/boost_1_63_0.tar.gz)
4.   `cd <Your-Project-Folder>/node_modules/react-native/third-party/glog-0.3.5`
5.   ` ../../scripts/ios-configure-glog.sh`
6.  Xcode Clean & Build

# Original issue

  - [x] Review the documentation: https://facebook.github.io/react-native
  - [x] Search for existing issues: https://github.com/facebook/react-native/issues
  - [x] Use the latest React Native release: https://github.com/facebook/react-native/releases

## Environment

  React Native Environment Info:
    System:
      OS: macOS High Sierra 10.13.6
      CPU: x64 Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
      Memory: 103.23 MB / 4.00 GB
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 10.9.0 - ~/.nvm/versions/node/v10.9.0/bin/node
      npm: 6.4.0 - ~/.nvm/versions/node/v10.9.0/bin/npm
    SDKs:
      iOS SDK:
        Platforms: iOS 12.0, macOS 10.14, tvOS 12.0, watchOS 5.0
    IDEs:
      Xcode: 10.0/10L221o - /usr/bin/xcodebuild
    npmPackages:
      react: 16.4.1 => 16.4.1 
      react-native: 0.56.0 => 0.56.0 
    npmGlobalPackages:
      react-native-cli: 2.0.1

## Description
First build request after yarn or npm install does not succeed due to apparent timing of download, extract, and compile actions. Subsequent builds succeed until node_modules removed and yarn or npm install run again.

Similar to this closed issue:
https://github.com/facebook/react-native/issues/18982

This is not a problem in the local development environment as the work around is to simply rerun the build command. However, with remote build environments (Buddybuild, App Center, etc)  this becomes problematic.

## Reproducible Demo

Follow these steps:

0 - configure environment as per react-native info above
1 - from shell - react-native init DoubleTrouble
2 - from shell - cd DoubleTrouble
3 - from shell - react-native run-ios

Here is the tail end of output I received:

** BUILD FAILED **

The following build commands failed:

	CompileC /Users/mcclen/Developer/DoubleTrouble/ios/build/Build/Intermediates.noindex/React.build/Debug-iphonesimulator/double-conversion.build/Objects-normal/x86_64/strtod.o /Users/mcclen/Developer/DoubleTrouble/node_modules/react-native/third-party/double-conversion-1.1.5/src/strtod.cc normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler

(1 failure)

Installing build/Build/Products/Debug-iphonesimulator/DoubleTrouble.app
An error was encountered processing the command (domain=NSPOSIXErrorDomain, code=22):
Failed to install the requested application
The bundle identifier of the application could not be determined.
Ensure that the application's Info.plist contains a value for CFBundleIdentifier.
Print: Entry, ":CFBundleIdentifier", Does Not Exist

Command failed: /usr/libexec/PlistBuddy -c Print:CFBundleIdentifier build/Build/Products/Debug-iphonesimulator/DoubleTrouble.app/Info.plist
Print: Entry, ":CFBundleIdentifier", Does Not Exist

### PR DESCRIPTION
**Pull request by mmccartney**
Fixes #20774 

The new Xcode build system uses parallel execution to run build steps that don't have an obvious dependency.  Our Xcode project was written with the assumption that the **Install Third Party** build step is run _before_ compiling the `third-party` libraries.  To address this issue, this PR adds dependency information to the project to teach Xcode that `ios-install-third-party.sh` is generating the files under `third-party`.  With this additional information, Xcode correctly waits for `ios-install-third-party.sh` to finish before advancing to the compile step.

In addition to the Xcode project changes, I had to make some changes to the script `ios-install-third-party.sh` so that
1. it would always execute the `ios-configure-glog.sh` script regardless of how it was invoked
2. it would always install the libraries even if Xcode had partially created the tree or if a previous install was interrupted

Test Plan:
----------
1. Under Xcode 9 and Xcode 10, delete the `third-party` tree and build the `React` project
1. Under Xcode 9 and Xcode 10, delete the `third-party` tree and build the `RNTester` project
1. Create and run a sample project (`react-native init AwesomeProject && cd AwesomeProject && react-native run-ios`)

Release Notes:
--------------
[IOS] [BUGFIX] [third-party] - Fix build issue using new Xcode 10.0 build system

### CODE REVIEW

**Comment by kelset** [comment]
Hey thanks a lot for this PR, I'm going to test it asap!

**Comment by radex** [comment]
I *think* the root of the problem is that Xcode 10 is trying to optimize build by parallelizing build steps, and it doesn't understand that the third-party install script is what creates files in the third-party. So it tries to go ahead with compile steps and fails.

captura de pantalla 2018-10-04 a las 14 36 28

I think the output files settings need to be used to fix this, but I haven't researched this fully yet ;)

**Comment by radex** [comment]
btw @mmccartney @kelset, I just did this:

captura de pantalla 2018-10-04 a las 14 49 59

and I'm 90% sure it works. I fully cleaned build artifacts, and ran a build from CLI, not Xcode (steps that would trigger the issue every time), and it builds successfully. Unless there's some weird non-determinism going on, this should be a good fix

**Comment by mmccartney** [comment]
Yeah, I thought there was some parallelization happening but wasn't sure.  I know it works if you run the script after `npm install` but before `react-native run-ios`.  How do people feel about a postinstall script run by `npm install` ?  I've never written one but would be willing to figure it out if people are open to it.

**Comment by kelset** [comment]
Go for it, if it's a viable solution we'll merge it 👍

**Comment by radex** [comment]
If you go this route (and it makes sense — I think — to put a dependency management script in… dependency management :P), I'd recommend removing the "Install Third Party" build phase from the Xcode project, as it doesn't do anything anymore

**Comment by ikesyo** [comment]
For CocoaPods users that is not needed, so I'm not for the `postinstall`.

**Comment by mmccartney** [comment]
> For CocoaPods users that is not needed, so I'm not for the `postinstall` (useless time consumption).

@ikesyo Are you saying that CocoaPods users are able to build react-native without the `third-party` artifacts?

**Comment by ikesyo** [comment]
Those are handled with https://github.com/facebook/react-native/tree/master/third-party-podspecs

http://facebook.github.io/react-native/docs/integration-with-existing-apps#configuring-cocoapods-dependencies

```ruby
  # Third party deps podspec link
  pod 'DoubleConversion', :podspec => '../node_modules/react-native/third-party-podspecs/DoubleConversion.podspec'
  pod 'glog', :podspec => '../node_modules/react-native/third-party-podspecs/glog.podspec'
  pod 'Folly', :podspec => '../node_modules/react-native/third-party-podspecs/Folly.podspec'
```

**Comment by mmccartney** [comment]
@radex it seems to me the *Install Third Party* step needs to be moved up to the `third-party` target.  I think my fix to the shell script itself is still needed though...  I'll give it a try

**Comment by mmccartney** [comment]
I didn't end up moving the *Install Third Party* build step up.  Instead I found that the build system wants a list of all outputs from the step in order to do dependency analysis.  So, for each third-party source file that is referenced within the Xcode project, I listed it in a new `third-party.xcfilelist` file which gives the build system the information it was missing before.

File: React/React.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -3,7 +3,7 @@
 	archiveVersion = 1;
 	classes = {
 	};
-	objectVersion = 46;
+	objectVersion = 51;
```
Comment (radex) [inline React/React.xcodeproj/project.pbxproj]: are you sure about this? isn't that Xcode 10-only-compatible mode?

**Comment by radex** [comment]
> I didn't end up moving the Install Third Party build step up. Instead I found that the build system wants a list of all outputs from the step in order to do dependency analysis. So, for each third-party source file that is referenced within the Xcode project, I listed it in a new third-party.xcfilelist file which gives the build system the information it was missing before.

I tried doing that, but it doesn't work correctly. Try deleting the third-party folder. Xcode will, at the beginning of the build, create each folder in the xcfilelist, and the install-third-party script will skip the download, because it thinks that it's already downloaded. We'd have to modify it to check if there are any files in the folder tree.

Also: I think `outputFileListPaths` makes the xcodeproj incompatible with Xcode 9.4 😭 — but you'd have to check

**Comment by kelset** [comment]
@radko93 I remember you having multiple Xcode instances, would you mind testing this PR to verify that it would be (or not) a BC for 9.x versions of Xcode? 🤔

**Comment by radko93** [comment]
My test plan for this (MacOS High Sierra, Xcode 9.2 (note that I also have Xcode 10 installed)):
1. Create a new RN project
2. Copy files changed in this PR by hand
3. Try running the project.

I can confirm I can build the app. @kelset

**Comment by radko93** [comment]
I will test this soon. If anyone else wants to do it than me please do.

**Comment by mmccartney** [comment]
@radex if `outputFileListPaths` is not compatible with Xcode 9, I think that `outputPaths` is and I could move all those files into the project.  I was just trying to use a medium that is easier to change when the third-party libraries change.

Unfortunately I don't have an Xcode 9 laptop setup right now.  I could probably get my hands on one next week.  Sorry I didn't think about supporting older Xcode.

@radko93 It may be good to hold off on testing for now.  I haven't found any documentation on exactly how to test this sort of thing and just discovered that `npm run-script build-ios-e2e` is failing with the following error (and others like it):

```
error: Multiple commands produce '/Users/mmccartney/react-native/RNTester/build/Build/Products/Release-iphonesimulator/libdouble-conversion.a':
1) Target 'double-conversion-tvOS' (project 'React') has a command with output '/Users/mmccartney/react-native/RNTester/build/Build/Products/Release-iphonesimulator/libdouble-conversion.a'
2) Target 'double-conversion' (project 'React') has a command with output '/Users/mmccartney/react-native/RNTester/build/Build/Products/Release-iphonesimulator/libdouble-conversion.a'
```

(of course, the same command fails on the _master_ branch due to missing `third-party` 😆 )

**Comment by radex** [comment]
>  if outputFileListPaths is not compatible with Xcode 9, I think that outputPaths is and I could move all those files into the project. 

You can do that, but it will be more difficult to automate when third-party libraries get updated. You could very easily make a script that writes an `.xcfilelist` file, but modifying an Xcode project file requires extra dependencies that I don't know if react-native already includes in its tooling

**Comment by mmccartney** [comment]
> You can do that, but it will be more difficult to automate when third-party libraries get updated

Yeah but honestly there are already so many file references inside the Xcode project that I don't think it matters.  Remember, I simply used all the references already in `project.pbxproj` to build the list.  There are a few other source references (e.g. `#include <folly/File.h>`) that I didn't bother including.

I'll go ahead and move them into `outputPaths` and revert the `objectVersion` changes (Xcode did that, not me 😉 ).

Can someone comment on whether `npm run-script build-ios-e2e` is supposed to work?  It obviously doesn't work on *master* but these tvOS vs iOS errors aren't making sense to me and make me wonder if it's unrelated...

**Comment by radex** [comment]
> Can someone comment on whether npm run-script build-ios-e2e is supposed to work? It obviously doesn't work on master but these tvOS vs iOS errors aren't making sense to me and make me wonder if it's unrelated...

I couldn't get this script to work on my machine, but a sure way to test whether Xcode 10 builds correctly when third-party is missing for me was:

1. open RNTester.xcodeproj in Xcode 10
2. Make sure to delete the third-party folder!
2. hit ⌘⇧K (clear build products)
3. hit ⌘B (build)

**Comment by mmccartney** [comment]
@radex Thanks for the instructions.  That worked for me after I ran `git clean -dfx` to get rid of previous artifacts (e.g. `xcuserdata` folders) that were causing it to fail.

I'm satisfied with my latest changes.  Please review and @radko93 if you are available to test, I think this is ready.

**Comment by mmccartney** [comment]
I saw some CI failures that seem unrelated to my changes so I just rebased from `master`.

**Comment by radko93** [comment]
@mmccartney I cannot test it on this branch since I can't do a complete `yarn install` because `detox` is broken (@kelset?). Then `postinstall` step is not being run

**Comment by radko93** [comment]
@mmccartney please rebase and let me know after it gets fixed and let me know afterwards

**Comment by mmccartney** [comment]
@radko93 I rebased

**Comment by mmccartney** [comment]
@radko93 I removed the `postinstall` step because of feedback from others.  Nothing to test there.  This PR now is a fix to the Xcode project so that it works with the new build system.  It _should_ work equally well with the old build system on Xcode 9.  I have tested with Legacy Build System as well as the New Build System using Xcode 10.  I will update the test plan (sorry about the oversight).

**Comment by mmccartney** [comment]
@kelset Do you know if the detox commit has landed?  I see a lot of new commits on `master` but nothing seems to be related to detox.  It looks like all the CI checks passed except AppVeyor which appears to have errors unrelated to my work?

**Comment by kelset** [comment]
> Do you know if the detox commit has landed? 

Not yet, @TheSavior is actively working on it atm.

> It looks like all the CI checks passed except AppVeyor which appears to have errors unrelated to my work?

Yeah you can ignore AppVeyor.

Aside from that, at the moment I don't have much time to review the final version of this PR, I'll ask a couple people to take a look asap.

**Pull request closed without merging**